### PR TITLE
Document Dart AST.

### PR DESCRIPTION
**Pull request by anyska**

### CODE REVIEW

File: src/dart/ast.rs [hunk 1]
```diff
@@ -385,53 +537,113 @@ impl Type {
...
     },
+    /// Zero-based ordered group of objects.
+    ///
+    /// `var list = [1, 2, 3]`
     List {
         const_: bool,
         element_ty: Option<Node<Type>>,
         elements: Vec<Node<Expr>>,
     },
+    /// An object that associates keys and values.
```
Comment (eddyb) [inline src/dart/ast.rs]: Same thing here (`Map<K, V>`).

File: src/dart/ast.rs [hunk 2]
```diff
@@ -385,53 +537,113 @@ impl Type {
...
+    ///
+    /// `var map = { 'k1' : 'v1', 'k2' : 'v2', 'k3' : 'v3'}`
     Map {
         const_: bool,
         kv_ty: Option<(Node<Type>, Node<Type>)>,
         kv: Vec<(Node<Expr>, Node<Expr>)>,
     },
+    /// A number (`1.2`).
```
Comment (eddyb) [inline src/dart/ast.rs]: "number literal".

File: src/dart/ast.rs [hunk 3]
```diff
@@ -385,53 +537,113 @@ impl Type {
...
     Number(Symbol),
+    /// Adjacent strings are implicitly concatenated
+    /// to form a single string literal.
```
Comment (eddyb) [inline src/dart/ast.rs]: I think "One or more string literals, implicitly concatenated to form a single \`String\`." is a bit more accurate.

File: src/dart/ast.rs [hunk 4]
```diff
@@ -512,16 +754,23 @@ impl Default for FnSig {
     }
 }
 
+/// The body of a function.
 #[derive(Debug)]
 pub enum FnBody {
+    /// The body is an expression preceded by a fat arrow (`=> expr`).
```
Comment (eddyb) [inline src/dart/ast.rs]: I say "fat" when talking to disambiguate but here "arrow" is enough. Should maybe mention that the expression's value is returned.

File: src/dart/ast.rs [hunk 5]
```diff
@@ -512,16 +754,23 @@ impl Default for FnSig {
...
     Arrow(Node<Expr>),
+    /// The body is a block statement (`{...}`).
     Block(Node<Statement>),
+    /// The body followed by native or native "..."
```
Comment (eddyb) [inline src/dart/ast.rs]: It's just \`native\` / \`native "..."\` (also, missing backticks).

File: src/dart/ast.rs [hunk 6]
```diff
@@ -512,16 +754,23 @@ impl Default for FnSig {
...
     Native(Option<StringLiteral>),
 }
 
+/// Optional parameters that can be either positional or named.
 #[derive(Copy, Clone, PartialEq, Eq, Debug)]
 pub enum OptionalArgKind {
+    /// Positional arguments.
```
Comment (eddyb) [inline src/dart/ast.rs]: \`[...]\`

File: src/dart/ast.rs [hunk 7]
```diff
@@ -512,16 +754,23 @@ impl Default for FnSig {
...
     Positional,
+    /// Named arguments.
```
Comment (eddyb) [inline src/dart/ast.rs]: \`{...}\`

File: src/dart/ast.rs [hunk 8]
```diff
@@ -512,16 +754,23 @@ impl Default for FnSig {
...
 pub enum FnBody {
+    /// The body is an expression preceded by a fat arrow (`=> expr`).
     Arrow(Node<Expr>),
+    /// The body is a block statement (`{...}`).
     Block(Node<Statement>),
+    /// The body followed by native or native "..."
     Native(Option<StringLiteral>),
 }
 
+/// Optional parameters that can be either positional or named.
```
Comment (eddyb) [inline src/dart/ast.rs]: "The style of optional arguments, if any."

File: src/dart/ast.rs [hunk 9]
```diff
@@ -531,13 +780,21 @@ impl Default for OptionalArgKind {
     }
 }
 
+/// Formal argument.
```
Comment (eddyb) [inline src/dart/ast.rs]: "Formal function argument definition."

File: src/dart/ast.rs [hunk 10]
```diff
@@ -557,71 +814,173 @@ impl ArgDef {
...
 }
 
+/// Variable qualifier.
 #[derive(Copy, Clone, PartialEq, Eq, Debug)]
 pub enum FinalConstVar {
+    /// A final variable can only be set once (`final name = 'Dart'`).
     Final,
+    /// A constant variable is a compile-time constant (`const length = 420`).
     Const,
+    /// Declares a variable without specifying its type (`var a = 0`).
```
Comment (eddyb) [inline src/dart/ast.rs]: Not necessarily! The type being unspecified is represented by `Type::Infer`, we've tested and apparently having both `var` and a type name (i.e. `var Type name;`) is valid.

File: src/dart/ast.rs [hunk 11]
```diff
@@ -557,71 +814,173 @@ impl ArgDef {
...
 pub enum Statement {
+    /// Comments that describe the item.
     Comments(Vec<Span>, Option<Node<Statement>>),
+    /// A block of statements (`{ ... }`).
     Block(Vec<Node<Statement>>),
+    /// Statement to declare variables.
+    ///
+    /// `Type var_name`
     Vars(VarType, Vec<Node<VarDef>>),
+    /// Statement containing a function.
```
Comment (eddyb) [inline src/dart/ast.rs]: "containing" -> "defining"

File: src/dart/ast.rs [hunk 12]
```diff
@@ -557,71 +814,173 @@ impl ArgDef {
...
+    /// The finally clause that contains a block statement.
+    ///
+    /// try {...} catch (e) {...} finally {...}
     pub block: Node<Statement>,
 }
 
+/// A statement.
 #[derive(Debug)]
 pub enum Statement {
+    /// Comments that describe the item.
```
Comment (eddyb) [inline src/dart/ast.rs]: "item" -> "statement"

File: src/dart/ast.rs [hunk 13]
```diff
@@ -557,71 +814,173 @@ impl ArgDef {
...
 }
 
+/// The catch clause of exception handlers.
 #[derive(Debug)]
 pub struct CatchPart {
+    /// The exception object.
+    ///
+    /// {...} catch(e) { print($e); }
     pub exception: Node<VarDef>,
+    /// The stacktrace object associated to the exception object.
```
Comment (eddyb) [inline src/dart/ast.rs]: Should indicate that this is optional somehow.

File: src/dart/ast.rs [hunk 14]
```diff
@@ -212,34 +301,53 @@ impl FnName {
...
 }
 
+/// Optionally fully-qualified path.
 #[derive(Debug)]
 pub struct Qualified {
+    /// The object to look up the name in.
     pub prefix: Option<Node<Qualified>>,
+    /// The name of the object.
     pub name: Symbol,
+    /// The type parameters of the object (`List<int> Map<K,V>`).
```
Comment (eddyb) [inline src/dart/ast.rs]: These are two examples, so they should be in separate code snippets, with a comma between them.

File: src/dart/ast.rs [hunk 15]
```diff
@@ -92,17 +97,29 @@ impl Module {
...
+/// An item.
 #[derive(Debug)]
 pub enum Item {
+    /// Provides a name to the library (`library name`).
     LibraryName { meta: Meta, path: Vec<Symbol> },
+    /// A directive that specifies how a namespace from one library is used
+    /// in the scope of another library (`import 'library'`).
     Import(Meta, Import),
+    /// A directive that makes the libraries used available to other libraries
+    /// should they import the current library (`export 'library'`).
```
[the remark this answers is not in the document]
Comment (anyska) [reply src/dart/ast.rs]: Users of *this* library? Doesn't sound quite right to me.

[the remark this answers is not in the document]
Comment (eddyb) [reply src/dart/ast.rs]: Ambiguity between "this" as "the library mentioned by the directive" and "this" as "the library we're in right now"?

[the remark this answers is not in the document]
Comment (anyska) [reply src/dart/ast.rs]: Yes, I'd rather use "current" instead of "this".

[the remark this answers is not in the document]
Comment (eddyb) [reply src/dart/ast.rs]: Sounds good.  It should also work for `Import` (even "current scope") and `LibraryName`.

File: src/dart/ast.rs [hunk 16]
```diff
@@ -121,57 +140,91 @@ pub enum Item {
...
     pub sig: FnSig,
+    /// The block of statements to be executed upon calling the function.
     pub body: Option<FnBody>,
 }
 
+/// A class member.
 #[derive(Debug)]
 pub enum ClassMember {
+    /// Comments that describe the class member.
     Comments(Vec<Span>),
```
[the remark this answers is not in the document]
Comment (anyska) [reply src/dart/ast.rs]: ??

File: src/dart/ast.rs [hunk 17]
```diff
@@ -385,53 +537,113 @@ impl Type {
...
     Closure(FnSig, FnBody),
+    /// A constructor call.
+    ///
+    /// `new Constructor()`
     New {
         const_: bool,
         path: Node<Qualified>,
         args: Args,
     },
+    /// Zero-based ordered group of objects.
```
[the remark this answers is not in the document]
Comment (anyska) [reply src/dart/ast.rs]: Drop the example as well?

File: src/dart/ast.rs (hunk 4 above)
Comment (anyska) [reply src/dart/ast.rs]: They use it in the Dart documentation as well, fwiw. ^_^

Comment (anyska) [reply src/dart/ast.rs]: "Maybe" is a bit lax.

File: src/dart/ast.rs (hunk 16 above)
[the remark this answers is not in the document]
Comment (eddyb) [reply src/dart/ast.rs]: If you rebase on master, it's already gone (@rovy previously forgot to remove it, did it today and I just merged the PR).

File: src/dart/ast.rs (hunk 4 above)
Comment (eddyb) [reply src/dart/ast.rs]: Oh then "fat arrow" is fine.

File: src/dart/ast.rs [hunk 18]
```diff
@@ -92,17 +97,29 @@ impl Module {
...
+/// A top-level item.
 #[derive(Debug)]
 pub enum Item {
+    /// Provides the name of the library (`library name;`).
     LibraryName { meta: Meta, path: Vec<Symbol> },
+    /// A directive that bring items from a library
+    /// in the scope of another library (`import 'library';`).
     Import(Meta, Import),
+    /// A directive that makes the items of a libraries available to
+    /// users of the current library (`export 'library'`).
```
Comment (eddyb) [inline src/dart/ast.rs]: Missing `;` in example.

Comment (eddyb) [inline src/dart/ast.rs]: Missing `;` in example.

File: src/dart/ast.rs [hunk 19]
```diff
@@ -92,17 +97,29 @@ impl Module {
...
     Export(Meta, StringLiteral, Vec<ImportFilter>),
+    /// A directive that specifies where a file that should be incorporated
+    /// into the current library is located (`part 'library'`).
```
Comment (eddyb) [inline src/dart/ast.rs]: Missing `;` in example.

File: src/dart/ast.rs [hunk 20]
```diff
@@ -92,17 +97,29 @@ impl Module {
     }
 }
 
+/// A top-level item.
 #[derive(Debug)]
 pub enum Item {
+    /// Provides the name of the library (`library name;`).
     LibraryName { meta: Meta, path: Vec<Symbol> },
+    /// A directive that bring items from a library
+    /// in the scope of another library (`import 'library';`).
```
Comment (eddyb) [inline src/dart/ast.rs]: Shouldn't "another" be "current", like for `export`?

File: src/dart/ast.rs (hunk 18 above)
Comment (eddyb) [reply src/dart/ast.rs]: "a libraries"

File: src/dart/ast.rs [hunk 21]
```diff
@@ -92,17 +97,29 @@ impl Module {
...
     Export(Meta, StringLiteral, Vec<ImportFilter>),
+    /// A directive that specifies where a file that should be incorporated
+    /// into the current library is located (`part 'library'`).
     Part {
         meta: Meta,
         uri: StringLiteral,
         module: Node<Module>,
     },
+    /// A directive that specifies that the current file is incorporated
+    /// in a library (`part of 'library'`).
```
Comment (eddyb) [inline src/dart/ast.rs]: Missing `;` in example.

File: src/dart/ast.rs [hunk 22]
```diff
@@ -92,17 +97,29 @@ impl Module {
...
     PartOf { meta: Meta, path: Vec<Symbol> },
+    /// A class defines the form and behaviour of a set of objects
+    /// that are its instances (`class { ... }`).
```
Comment (eddyb) [inline src/dart/ast.rs]: Missing name in example.

File: src/dart/ast.rs [hunk 23]
```diff
@@ -113,6 +130,8 @@ pub enum Item {
         interfaces: Vec<Node<Qualified>>,
         members: Vec<Node<ClassMember>>,
     },
+    /// A mixin describes the difference between a class and its superclass
```
Comment (eddyb) [inline src/dart/ast.rs]: "the difference" is still suboptimal, how about "A mixin class combines the members of multiple classes."?

File: src/dart/ast.rs [hunk 24]
```diff
@@ -121,56 +140,90 @@ pub enum Item {
...
+    /// Enumerated types are a special kind of class used to represent
+    /// a fixed set of constant values.
+    ///
+    /// (`enum Type { integer, boolean, string }`).
     Enum {
         meta: Meta,
         name: Symbol,
         values: Vec<(Meta, Symbol)>,
     },
+    /// A type alias defines a name for a given type
```
Comment (eddyb) [inline src/dart/ast.rs]: "a *new* name"?

File: src/dart/ast.rs [hunk 25]
```diff
@@ -121,56 +140,90 @@ pub enum Item {
...
         generics: Vec<Node<TypeParameter>>,
         ty: Node<Type>,
     },
+    /// Function definition (`example(x) { return x }`).
     Function {
         meta: Meta,
         external: bool,
         function: Node<Function>,
     },
+    /// Global variables (`var example = 123`).
```
Comment (eddyb) [inline src/dart/ast.rs]: Missing `;` in example.

File: src/dart/ast.rs [hunk 26]
```diff
@@ -121,56 +140,90 @@ pub enum Item {
...
     Vars(Meta, VarType, Vec<Node<VarDef>>),
 }
 
+/// Specifies which items of the imported library to be shown/hidden.
 #[derive(Clone, Debug)]
 pub struct ImportFilter {
+    /// Specifies the action to be applied to the items specified.
     pub hide: bool,
+    /// Specifies the items of the library.
```
Comment (eddyb) [inline src/dart/ast.rs]: Maybe add "to show/hide" at the end?

File: src/dart/ast.rs [hunk 27]
```diff
@@ -121,56 +140,90 @@ pub enum Item {
...
     pub names: Vec<Symbol>,
 }
 
+/// Specifies a library to be used in the scope of the current library.
+///
+/// `import 'dart:specifier';`
```
Comment (eddyb) [inline src/dart/ast.rs]: `dart:io` is probably a better example here since it's short and used in real code.

File: src/dart/ast.rs [hunk 28]
```diff
@@ -179,7 +232,13 @@ pub enum ClassMember {
         initializers: Vec<ConstructorInitializer>,
         function_body: Option<FnBody>,
     },
+    /// A function that applies only to the object and provides behaviour.
+    ///
+    /// `printObject() { print(object_field) }`
     Method(Meta, Vec<MethodQualifiers>, Node<Function>),
+    /// Variables that describe the state of the class instances.
+    ///
+    /// `var object_field = 0`
```
Comment (eddyb) [inline src/dart/ast.rs]: `int counter = 0;` might work better?

File: src/dart/ast.rs [hunk 29]
```diff
@@ -288,21 +414,36 @@ impl BoolBinOp {
...
-        BitAnd,
-        BitXor,
-        BitOr,
-    }
+/// Value binary operators.
+#[derive(Copy, Clone, PartialEq, Eq, Debug)]
+pub enum ValueBinOp {
+    /// Conditional expression (`a ?? b`).
+    /// If a is non-null, returns its value;
+    /// otherwise, evaluates and returns the value of b.
```
Comment (eddyb) [inline src/dart/ast.rs]: Missing backticks around mentions of `a` and `b`.

File: src/dart/ast.rs [hunk 30]
```diff
@@ -384,53 +536,113 @@ impl Type {
...
     Conditional(Node<Expr>, Node<Expr>, Node<Expr>),
+    /// An expression that checks the type (`var is type`).
     Is(Node<Expr>, Node<Type>),
+    /// A negated expression that checks the type (`var !is type`).
     IsNot(Node<Expr>, Node<Type>),
+    /// An expression that acts as a cast (`x as Type`).
     As(Node<Expr>, Node<Type>),
+    /// A sequence of operator, method, getter or setter invocations.
+    ///
+    /// `expr.suffix`
```
Comment (eddyb) [inline src/dart/ast.rs]: `expr.field` and `expr[idx]` are probably good examples.

File: src/dart/ast.rs [hunk 31]
```diff
@@ -384,53 +536,113 @@ impl Type {
...
     Suffix(Node<Expr>, Suffix),
+    /// An identifier (`x`).
     Identifier(Symbol),
+    /// A function object that has access to variables in its lexical scope,
+    /// even when the function is used outside of its original scope.
+    ///
+    ///`(a, b) => a + b`
```
Comment (eddyb) [inline src/dart/ast.rs]: Missing space after `///`.

File: src/dart/ast.rs [hunk 32]
```diff
@@ -384,53 +536,113 @@ impl Type {
...
     pub raw: bool,
+    /// Triple matching quotes allow multiline strings.
+    ///
+    /// `'''`
+    ///
+    /// `Multi-line strings`
+    ///
+    /// `like this one`
+    ///
+    /// `are allowed through triple quotes.'''`
```
Comment (eddyb) [inline src/dart/ast.rs]: I think this should be \`\`\` before and after, with no empty lines in between.

File: src/dart/ast.rs [hunk 33]
```diff
@@ -441,36 +653,58 @@ impl StringLiteral {
     }
 }
 
+/// Non-unary postfix operations on expressions.
 #[derive(Debug)]
 pub enum Suffix {
+    /// Indexing suffix (`..children[0]`).
```
Comment (eddyb) [inline src/dart/ast.rs]: `expr[0]` (or `expr[idx]`)

File: src/dart/ast.rs [hunk 34]
```diff
@@ -441,36 +653,58 @@ impl StringLiteral {
...
     Index(Node<Expr>),
+    /// Field suffix (`..name`).
```
Comment (eddyb) [inline src/dart/ast.rs]: `expr.name`

File: src/dart/ast.rs [hunk 35]
```diff
@@ -441,36 +653,58 @@ impl StringLiteral {
...
     Field(Symbol),
+    /// Gets the field if the expression returns a valid object (`?..name`).
```
Comment (eddyb) [inline src/dart/ast.rs]: `expr.?name` (I think? may need double-checking)

File: src/dart/ast.rs [hunk 36]
```diff
@@ -441,36 +653,58 @@ impl StringLiteral {
...
     FieldIfNotNull(Symbol),
+    /// Performs a function call on the object (`..name.toUpperCase()`).
```
Comment (eddyb) [inline src/dart/ast.rs]: `expr(a, b, c)`

File: src/dart/ast.rs [hunk 37]
```diff
@@ -486,15 +720,23 @@ impl MetaItem {
...
     pub return_type: Node<Type>,
+    /// The list of required parameters.
     pub required: Vec<ArgDef>,
+    /// The list of optional parameters.
     pub optional: Vec<ArgDef>,
+    /// Optional parameters to be bound.
     pub optional_kind: OptionalArgKind,
+    /// Asynchronous function denoted through async or async*.
     pub async: bool,
+    /// Generator function denoted through sync* or async*.
```
Comment (eddyb) [inline src/dart/ast.rs]: The last two comments should have backticks around the `async` & friends.

File: src/dart/ast.rs [hunk 38]
```diff
@@ -511,16 +753,24 @@ impl Default for FnSig {
...
 
+/// The body of a function.
 #[derive(Debug)]
 pub enum FnBody {
+    /// The body is an expression preceded by an arrow and
+    /// it returns its value (`=> expr`).
     Arrow(Node<Expr>),
+    /// The body is a block statement (`{...}`).
     Block(Node<Statement>),
+    /// The body is `native` / `native "..."`
```
Comment (eddyb) [inline src/dart/ast.rs]: Missing `.`.

File: src/dart/ast.rs [hunk 39]
```diff
@@ -530,13 +780,21 @@ impl Default for OptionalArgKind {
...
+    /// The metadata of the argument.
     pub meta: Meta,
+    /// Allows overriding the type with a subtype (`covariant T x`).
     pub covariant: bool,
+    /// The type of the argument.
     pub ty: VarType,
+    /// Qualifies the argument as field, used in constructors (`this.x`).
     pub field: bool,
+    /// Specifies if the default value is given through `=` instead of `:`.
+    /// `(int x = 5) instead of (int x: 5)`
```
Comment (eddyb) [inline src/dart/ast.rs]: I think there should be more backticks, i.e.:
\`(int x = 5)\` instead of \`(int x: 5)\`

File: src/dart/ast.rs [hunk 40]
```diff
@@ -556,71 +814,175 @@ impl ArgDef {
...
 }
 
+/// Variable qualifier.
 #[derive(Copy, Clone, PartialEq, Eq, Debug)]
 pub enum FinalConstVar {
+    /// A final variable can only be set once (`final name = 'Dart'`).
     Final,
+    /// A constant variable is a compile-time constant (`const length = 420`).
     Const,
+    /// Declares a variable without necessarily specifying its type (`var a = 0`).
```
Comment (eddyb) [inline src/dart/ast.rs]: The last 3 comments should have `;`.

File: src/dart/ast.rs [hunk 41]
```diff
@@ -556,71 +814,175 @@ impl ArgDef {
...
     Var,
 }
 
+/// Var definition.
 #[derive(Debug)]
 pub struct VarDef {
+    /// The name assigned to the variable (`var x`).
     pub name: Symbol,
+    /// The initial value given to the variable upon declaration (`var x = 0`).
```
Comment (eddyb) [inline src/dart/ast.rs]: The last 3 comments should have `;`.

File: src/dart/ast.rs [hunk 42]
```diff
@@ -556,71 +814,175 @@ impl ArgDef {
...
     pub init: Option<Node<Expr>>,
 }
 
+/// Control flow statement that iterates over a condition or collection.
 #[derive(Debug)]
 pub enum ForLoop {
+    /// Standard form.
+    ///
+    /// `for(var v=0; condition; expr) {...}`
```
Comment (eddyb) [inline src/dart/ast.rs]: `v=0` should have spaces around `=`.

File: src/dart/ast.rs [hunk 43]
```diff
@@ -556,71 +814,175 @@ impl ArgDef {
...
     CLike(Node<Statement>, Option<Node<Expr>>, Vec<Node<Expr>>),
+    /// A for loop that iterates over a collection.
+    ///
+    /// `for(x in collection) {...}`
     In(Symbol, Node<Expr>),
+    /// A for loop that iterates over a collection while also
+    /// declaring the auxiliary variable.
```
Comment (eddyb) [inline src/dart/ast.rs]: "the variable for the iteration element."?

File: src/dart/ast.rs [hunk 44]
```diff
@@ -556,71 +814,175 @@ impl ArgDef {
...
+    /// The actions to be executed should the case match.
     pub statements: Vec<Node<Statement>>,
 }
 
+/// The catch clause of exception handlers.
 #[derive(Debug)]
 pub struct CatchPart {
+    /// The exception object.
+    ///
+    /// {...} catch(e) { print($e); }
```
Comment (eddyb) [inline src/dart/ast.rs]: Missing backticks, here and in 3 more comments just below.

File: src/dart/ast.rs (hunk 26 above)
Comment (anyska) [reply src/dart/ast.rs]: Again too lax.

File: src/dart/ast.rs (hunk 29 above)
Comment (anyska) [reply src/dart/ast.rs]: Isn't the whole expression the example? Why would `a` and `b` be separately backticked?

**Pull request merged**